text/3553-cargo-sbom.md

```diff
@@ -0,0 +1,121 @@
+
+New government initiatives aimed at improving the security of the software supply chain such as the US "Executive Order 14028: Improving the Nation's Cybersecurity" or the EU "Cyber Resilience Act" require a Software Bill of Materials. Generating accurate SBOMs with Cargo is currently difficult because, depending on target selection or activated features, the dependencies may be different.
+
+For workspaces that generate multiple compiled artifacts, each artifact may have different dependencies referenced. Existing tools (see prior art section) attempt to approximate the correct dependency set, however precise dependency information for each compiled artifact is difficult without built-in Cargo support. Generating the SBOM information at the same time as the compiled artifact allows precise dependency information to be emitted for each compiled artifact.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The generation of SBOM information is controlled by Cargo's configuration. To enable SBOM generation, set the following configuration:
+
+```toml
+[build]
+sbom = true
+```
+
+Or use the environment variable `CARGO_BUILD_SBOM=true`.
+
+If enabled, an SBOM file will be placed next to each compiled artifact for `bin`, `staticlib`, `cdylib` crate types in the `target` directory with the name `<artifact>.cargo-sbom.json`. The SBOM will contain information about dependencies used to build the compiled artifact.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The SBOM file generated by Cargo is *not* intended as a final SBOM artifact, but rather a precursor. Post-processing tooling can use the information produced here as part of building a final SBOM.
+
+The SBOM file will be written to disk before `rustc` is executed for the each artifact. This enables [`RUSTC_WORKSPACE_WRAPPER`](https://doc.rust-lang.org/cargo/reference/config.html#buildrustc-workspace-wrapper) to point at a program that can utilize the SBOM file to embed information into the binary if desired. The environment variable `CARGO_SBOM_PATH` will be set to the full path to the SBOM file.
+
+## Format
+The format will use JSON, but the exact format is not specified in this RFC. Additional fields can be added as needed. The JSON will include a `format-version` in case breaking changes are necessary.
+
+### Resolved Dependency Tree
+The SBOM file will include the following information for each crate. Note that a crate refers to a [single build unit (library or executable)](https://doc.rust-lang.org/cargo/appendix/glossary.html#crate).
+A [package](https://doc.rust-lang.org/cargo/appendix/glossary.html#package) may contain multiple crates.
+- Package ID
+- Name
+- Version
+- Source (registry / git / path etc.)
+- Checksum (if available)
+- Dependencies
```

Comment:
Note that there are two ways of looking at dependencies: what each package needs, and the final resolved graph.

For example, if one package depends on `rand` with `features = ["std", "getrandom"]`, and another with `features = ["std", "simd_support"]`, the final resolved features will be `["std", "getrandom", "simd_support"]`. Depending on the use case you may need either or both representations (direct package dependencies and the resolved graph).

`cargo metadata` [exposes both](https://doc.rust-lang.org/cargo/commands/cargo-metadata.html) (under "packages" and "resolve" fields), but inaccurately: 
 - https://github.com/rust-lang/cargo/issues/7754
 - Under resolver v2, it [conflates the normal and build-dependency trees](https://rust-lang.github.io/rfcs/2957-cargo-features2.html#cargo-metadata)

I think it would be best for the SBOM to also expose both, accurately this time.